src/librustc/infer/error_reporting/named_anon_conflict.rs

```diff
@@ -0,0 +1,115 @@
+    // by that of the named region.
+    pub fn find_arg_with_anonymous_region(&self,
+                                          anon_region: Region<'tcx>,
+                                          named_region: Region<'tcx>)
+                                          -> Option<(&hir::Arg, ty::Ty<'tcx>)> {
+
+        match *anon_region {
+            ty::ReFree(ref free_region) => {
+
+                let id = free_region.scope;
+                let node_id = self.tcx.hir.as_local_node_id(id).unwrap();
+                let body_id = self.tcx.hir.maybe_body_owned_by(node_id).unwrap();
+
+                let body = self.tcx.hir.body(body_id);
+                body.arguments
+                    .iter()
+                    .filter_map(|arg| if let Some(tables) = self.in_progress_tables {
+                                    let ty = tables.borrow().node_id_to_type(arg.id);
+                                    let mut found_anon_region = false;
+                                    let new_arg_ty = self.tcx
+                                        .fold_regions(&ty,
+                                                      &mut false,
+                                                      |r, _| if *r == *anon_region {
+                                                          found_anon_region = true;
+                                                          named_region
+                                                      } else {
+                                                          r
+                                                      });
+                                    if found_anon_region {
+                                        return Some((arg, new_arg_ty));
+                                    } else {
+                                        None
+                                    }
+                                } else {
+                                    None
+                                })
+                    .next()
+            }
+            _ => None,
+        }
+
+    }
+
+    // This method generates the error message for the case when
+    // the function arguments consist of a named region and an anonymous
+    // region and corresponds to `ConcreteFailure(..)`
+    pub fn report_named_anon_conflict(&self, error: &RegionResolutionError<'tcx>) -> bool {
+
+        let (span, sub, sup) = match *error {
+            ConcreteFailure(ref origin, sub, sup) => (origin.span(), sub, sup),
+            _ => return false, // inapplicable
+        };
+
+        let (named, (var, new_ty)) =
+            if self.is_named_region(sub) && self.is_anonymous_region(sup) {
+                (sub, self.find_arg_with_anonymous_region(sup, sub).unwrap())
+            } else if self.is_named_region(sup) && self.is_anonymous_region(sub) {
+                (sup, self.find_arg_with_anonymous_region(sub, sup).unwrap())
+            } else {
+                return false; // inapplicable
+            };
```

Comment:
Somewhere in here I think we should get the scope of the anonymous region and check that it is defined on something suitable. Something like:

```
let scope_node_id = self.tcx.hir.as_local_node_id(scope_def_id).unwrap();
match self.tcx.hir.find(scope_node_id) {
  Some(hir_map::NodeItem(..)) |
  Some(hir_map::NodeTraitItem(..)) => { /* proceed ahead */ }
  Some(hir_map::NodeImplItem(..)) => {
    if tcx.impl_trait_ref(scope_def_id).is_some() {
      // For now, we do not try to target impls of traits. This is
      // because this message is going to suggest that the user
      // change the fn signature, but they may not be free to do so,
      // since the signature must match the trait.
      //
      // FIXME(#42706) -- in some cases, we could do better here.
      return false;
    }
  }
  _ => return false, // inapplicable; we target only top-level functions, not closures
}
```

This code currently lives in `is_anonymous_region`.

Alternatively, we could do the check in that function, but then let's move it into this file (it's now rather specific to this error message) and give it a different name, like `is_suitable_anonymous_region()`. That's maybe the best path.